pythonforandroid/recipe.py

```diff
@@ -583,6 +583,8 @@ def get_recipe(cls, name, ctx):
         if not hasattr(cls, "recipes"):
             cls.recipes = {}
         if name in cls.recipes:
+            if ctx is not None:
+                cls.recipes[name].ctx = ctx
```

Comment:
That really gets me curious :thinking:
Why would we need to override the context of the cached recipe? It should already be there via this https://github.com/kivy/python-for-android/blob/v2019.07.08/pythonforandroid/recipe.py#L610 right?

Reply:
Well the fishiness is that the code is written under the assumption just one context will ever be relevant which obviously isn't true at least for the tests, so the whole static caching may not be an ideal design. But for regular program execution of p4a it probably won't matter, but sure it is pretty weird in its inherent assumptions - not that this is really caused by this pull request

Reply:
OK then if it's the test that are misbehaving, should the tests be updated?
Anyway that looked properly weird to me so I checked out the branch and tried removing that lines. Tests are passing just fine

Reply:
@AndreMiras it depends on the test execution order I think, and if you just look at the tests it is obvious it is inherently, necessarily going to blow up. Look in `tests/test_recipe.py`, multiple tests construct a `Context()` more than once and these contexts are used for obtaining a recipe, so obviously if any ever obtain the same recipe you already got the disaster spelled out (because now you can possibly get the recipe attached with the wrong context of a previous test, and that can introduce all sorts of subtle wrong behavior)

Fixing the tests from my first conclusions doesn't make too much sense, because why even have the `ctx` parameter for getting a recipe if it's ignored? I think the bandaid most reasonable fix is what I added here, and the long-term fix might be to rethink this cached recipe approach in different ways (e.g. cache it with a dictionary mapping to context with per-context-and-recipe-pair singleton rather than global per-recipe singleton)

Reply:
OK I think I understand what you mean. But then in this case I still feel we're increasing the technical debt (if any) by adding this line that just feels unnatural to me :confused: 
Unit tests should be definition be isolated and not leave the context modified from one to another unless we specifically want to. And we should not put a workaround in the code to fix this misbehaviour. I don't think static attributes is anything new in Python or in other languages, same things like with singleton and stuff. When we have one we need to carefully write tests to not leave it in a state we don't want to. I know I'm completely guilty of writing some of such tests, thinking "it's fine I'll address when it impacts us"

Reply:
Just to emphasize this again, I think it is wrong to fix this in the tests. `get_recipe` has a context/`ctx` parameter, it is completely counter-intuitive and somewhat nonsensical to me that it is ever ignored. So if there is any fix I think it has to be inside `get_recipe`

I'm open to dropping this from the pull request (I put it on because I saw spurious test failures) but I would expect us to see weird errors down the line unless we do something about it. I agree this bandaid is not the nicest way to fix it, but breaking up the singleton design entirely also could have tons of implications and break a lot of things when done thoughtlessly, so I don't think a good solution will come quickly.

I'm against "fixing" anything in the tests themselves since again, I think this is just hiding the design problem of `get_recipe()` here and a worse technical debt than this workaround

Reply:
Very sad to ear :cry: 
In the order of preference I personally have (the first being my preferred solution):
1) of course a proper fix by design if you think it's a design flow (but as you said that would involve a lot of changes)
2) introduce "work-around" in the tests (I've implemented an example in https://github.com/kivy/python-for-android/pull/1948)
3) introduce "work-around" in the code (what you're currently implementing)

Anyway let's assume we keep your implementation of `3)`. Then could you please write a dedicated and comprehensive test (and docstring) showing this is needed and avoiding later regressions. I reckon the test should be fairly easy to write I think.
Then two thing come to my mind.
1) with your workaround, do we still need this line then https://github.com/kivy/python-for-android/blob/v2019.07.08/pythonforandroid/recipe.py#L610 ?
2) why in your workaround `None` is treated differently? Let's cover that case as well

Reply:
I think a work-around in a test is just masking it, so I really don't like the idea at all. However what you proposed as workaround actually seems like a useful long-term stability measure so it doesn't seem like the worst idea - just not good for dealing with this actual issue

I think we need to fix it, I'm just not sure how yet. I'll sleep about it and see if I can think of what might be the best way to address this :+1:

Reply:
(oh and I'll also back this "fix" out of this pull request here unless you disagree, since this is growing way too big/unrelated and needs more thought. I suggest we should handle it separately)

Reply:
Yes I would prefer to not have any of theses workaround or fixes in the PR. But if we do, then I would like them properly unit tested